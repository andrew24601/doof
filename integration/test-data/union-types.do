// Test union types and type narrowing

// Union type variable

// Type narrowing with typeof

// Change the value and test again

// Function that accepts union type

function processValue(input: int | string | bool): string {
    if (input is int) {
        return "N" + input;
    } else if (input is string) {
        return "S" + input;
    } else {
        return "B" + (input ? "T" : "F");
    }
}

function main(): int {
    let result = "";
    let value: int | string = 42;
    if (value is int) {
        result = result + "num:" + value + "|";
    } else {
        result = result + "str:" + value + "|";
    }
    value = "hello";
    if (value is int) {
        result = result + "num:" + value + "|";
    } else {
        result = result + "str:" + value + "|";
    }
    result = result + processValue(123) + "|";
    result = result + processValue("test") + "|";
    result = result + processValue(true) + "|";
    result = result + processValue(false);
    println(result);
    return 0;
}