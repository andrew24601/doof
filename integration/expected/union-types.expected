num:42|str:hello|N123|Stest|BT|BF